### PR TITLE
[IMP] core: remove @api.returns and move call_kw to odoo.service.model

### PR DESCRIPTION
**Pull request by kmagusiak**
Description of the issue/feature this PR addresses:

`call_kw` is a function that is specific to external calls and should be in the model service.

The `@api.returns` decorator logically gives two signatures to a single function. In that case, it is better to have 2 different functions: for testing and conceptually to avoid having a function which result type depends on the context. Moreover, from an architecture point of view, the type of the result for external calls should be handled by the controller/service.

Only some of the `mail` functions really used this feature and removing it simplifies the api module.

Current behavior before PR:
- `call_kw` inside `odoo.api`
- `@api.returns` not often used, basically to return a record.id depending on the context of the call

Desired behavior after PR is merged:
- `call_kw` inside `odoo.service.model` (where it makes sense with other dispatch functions)
- functions have one return type

odoo/enterprise#71424
odoo/documentation#11156
task-4155860

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by tde-banana-odoo** [comment]
:+1:

File: addons/mail/models/discuss/discuss_channel.py [hunk 1]
```diff
@@ -943,9 +942,11 @@ def _to_store(self, store: Store):
...
             store.add(channel, info)
 
-    # User methods
     @api.model
-    @api.returns('self', lambda channels: Store(channels).get_result())
+    def channel_get_store(self, partners_to, pin=True, force_open=False):
+        return Store(self.channel_get(partners_to, pin, force_open)).get_result()
+
+    @api.model
     def channel_get(self, partners_to, pin=True, force_open=False):
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply addons/mail/models/discuss/discuss_channel.py:950]: True, they probably could become private, but I would have to check usages/overwrites.
@tde-banana-odoo Do you think I should rewrite all 3 methods in that files to the likes of `channel_get` becomes `_channel_get` and `channel_get_store` becomes `channel_get`?

File: addons/mail/models/mail_thread.py [hunk 2]
```diff
@@ -2068,7 +2068,6 @@ def _get_customer_information(self):
     def _get_allowed_message_post_params(self):
         return {"attachment_ids", "body", "message_type", "partner_ids", "subtype_xmlid"}
 
-    @api.returns('mail.message', lambda value: value.id)
     def message_post(self, *,
```
[the remark this answers is not in the document]
Comment (kmagusiak) [reply addons/mail/models/mail_thread.py:2071]: For RPC users, they will get results closer to what is really happening in the model. If a specific return type is needed, you may create another function or implement it in a controller.
In the response, you get an array of ids; there was even some code I could simplify which was using regexes and perseInt.

For this case in particular, even the `mail` module uses a controller `/mail/message/post`. In my opinion, most RPC calls just want to post a message and don't need the resulting id of the message.

File: addons/mail/models/discuss/discuss_channel.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/discuss/discuss_channel.py:950]: Can't say for channel model, don't know the internals of store API. For mail obviously message_post is still an official stuff you can call :disguised_face:

File: addons/mail/models/mail_thread.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (tde-banana-odoo) [reply addons/mail/models/mail_thread.py:2071]: We should not prevent calls to message_post, it is considered as a valid "public" method. Controllers are more specific to discuss / chatter, but you can call message_post without having to rely on controllers.

[the remark this answers is not in the document]
Comment (kmagusiak) [reply addons/mail/models/mail_thread.py:2071]: You still can call it.

File: addons/mail/models/discuss/discuss_channel.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply addons/mail/models/discuss/discuss_channel.py]: @seb-odoo Which of the change you prefer? Having `channel_get/channel_get_store` or `_channel_get/channel_get` where the second one returns the result of the `Store` object?

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/models/discuss/discuss_channel.py]: I'd say I prefer to make the public one a route (not an extra model method) and make the model method private.

[the remark this answers is not in the document]
Comment (kmagusiak) [reply addons/mail/models/discuss/discuss_channel.py]: Ok, I will make the method private. To make it a controller method, there would be some more work required (not yet done).

**Review by rco-odoo — COMMENTED** [review]
> `call_kw` is a function that is specific to external calls and should be in the model service.

My motivation for having this function in `api.py` was that it is an API _adapter_.  And it is not specific to _external_ calls, since it is called by the XML file importer (as part of the implementation of tag `<function>`).  The latter is used when installing/upgrading modules, and I wouldn't call that an _external_ call...

File: addons/mail/models/discuss/discuss_channel.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmagusiak) [reply addons/mail/models/discuss/discuss_channel.py]: Separated into #189366

**Review by rco-odoo — APPROVED** [review]
@robodoo rebase-ff r+

**Comment by rco-odoo** [comment]
@robodoo override=ci/security

All the flagged code comes from code that has been moved.

**Pull request closed without merging**